File: app/src/test/java/igrek/songbook/service/songtree/SongTreeSorterTest.java
package igrek.songbook.service.songtree;

import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import igrek.songbook.persistence.songsdb.Song;
import igrek.songbook.persistence.songsdb.SongCategory;
import igrek.songbook.persistence.songsdb.SongCategoryType;
import igrek.songbook.persistence.songsdb.SongStatus;
import igrek.songbook.songselection.tree.SongTreeItem;
import igrek.songbook.songselection.tree.SongTreeSorter;

import static org.assertj.core.api.Java6Assertions.assertThat;

public class SongTreeSorterTest {
	
	private SongTreeSorter sorter = new SongTreeSorter();
	
	@Test
	public void test_songTreeItemSorter() {
		// empty list
		assertSortedOrdering(Arrays.asList(), Arrays.asList());
		
		SongTreeItem song51 = songItem("51");
		SongTreeItem songCma = songItem("Ćma barowa");
		SongTreeItem songJestTaki = songItem("jest taki samotny dom");
		SongTreeItem songJolkaJolka = songItem("Jolka, jolka");
		SongTreeItem songZzz = songItem("Żśźęćół");
		SongTreeItem categoryCustom = categoryItem(SongCategoryType.CUSTOM, "Custom");
		SongTreeItem categoryBudka = categoryItem("Budka suflera");
		SongTreeItem categoryCc = categoryItem("Ćć");
		SongTreeItem categoryPinkFloyd = categoryItem("Pink Floyd");
		SongTreeItem categoryOthers = categoryItem(SongCategoryType.OTHERS, "Others");
		
		// one song
		assertSortedOrdering(Arrays.asList(song51), Arrays.asList(song51));
		
		// simple songs
		assertSortedOrdering(Arrays.asList(songJolkaJolka, songJestTaki), Arrays.asList(songJestTaki, songJolkaJolka));
		
		// categories
		assertSortedOrdering(Arrays.asList(categoryBudka, categoryOthers, categoryCustom, categoryPinkFloyd, categoryCc), Arrays
				.asList(categoryCustom, categoryBudka, categoryCc, categoryPinkFloyd, categoryOthers));
		
		// all
		assertSortedOrdering(Arrays.asList(categoryBudka, songZzz, songCma, songJolkaJolka, song51, categoryPinkFloyd, categoryOthers, categoryCustom, songJestTaki, categoryCc), Arrays
				.asList(song51, songCma, songJestTaki, songJolkaJolka, songZzz, categoryCustom, categoryBudka, categoryCc, categoryPinkFloyd, categoryOthers));
	}
	
	private SongTreeItem songItem(String title) {
		return songItem(title, SongCategoryType.OTHERS, "others", false);
	}
	
	private SongTreeItem songItem(String title, String artist) {
		return songItem(title, SongCategoryType.ARTIST, artist, false);
	}
	
	private SongTreeItem songItem(String title, SongCategoryType categoryType, String categoryName, boolean custom) {
		SongCategory category = new SongCategory(categoryType.getId(), categoryType, categoryName, false, categoryName, new ArrayList<>());
		Song song = new Song(1, title, category, null, 1, 0, 0, custom, title, null, null, false, null, null, SongStatus.PUBLISHED, null, null, null, null, null, null, null, null);
		return SongTreeItem.Companion.song(song);
	}
	
	private SongTreeItem categoryItem(String artist) {
		return categoryItem(SongCategoryType.ARTIST, artist);
	}
	
	private SongTreeItem categoryItem(SongCategoryType categoryType, String categoryName) {
		SongCategory category = new SongCategory(categoryType.getId(), categoryType, categoryName, false, categoryName, new ArrayList<>());
		return SongTreeItem.Companion.category(category);
	}
	
	private void assertSortedOrdering(List<SongTreeItem> in, List<SongTreeItem> outExpected) {
		List<SongTreeItem> out = sorter.sort(in);
		assertThat(out).isEqualTo(outExpected);
	}
	
}

File: app/src/test/java/igrek/songbook/service/songtree/SongTreeSorterTest.kt
package igrek.songbook.service.songtree

import igrek.songbook.persistence.general.model.Category
import igrek.songbook.persistence.general.model.CategoryType
import igrek.songbook.persistence.general.model.Song
import igrek.songbook.persistence.general.model.SongStatus
import igrek.songbook.songselection.listview.items.SongTreeItem
import igrek.songbook.songselection.tree.SongTreeSorter
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class SongTreeSorterTest {

    private val sorter = SongTreeSorter()

    @Test
    fun test_songTreeItemSorter() {
        // empty list
        assertSortedOrdering(listOf(), listOf())

        val song51 = songItem("51")
        val songCma = songItem("Ćma barowa")
        val songJestTaki = songItem("jest taki samotny dom")
        val songJolkaJolka = songItem("Jolka, jolka")
        val songZzz = songItem("Żśźęćół")
        val categoryCustom = categoryItem(CategoryType.CUSTOM, "Custom")
        val categoryBudka = categoryItem("Budka suflera")
        val categoryCc = categoryItem("Ćć")
        val categoryPinkFloyd = categoryItem("Pink Floyd")
        val categoryOthers = categoryItem(CategoryType.OTHERS, "Others")

        // one song
        assertSortedOrdering(listOf(song51), listOf(song51))

        // simple songs
        assertSortedOrdering(listOf(songJolkaJolka, songJestTaki), listOf(songJestTaki, songJolkaJolka))

        // categories
        assertSortedOrdering(listOf(categoryBudka, categoryOthers, categoryCustom, categoryPinkFloyd, categoryCc), listOf(categoryCustom, categoryBudka, categoryCc, categoryPinkFloyd, categoryOthers))

        // all
        assertSortedOrdering(listOf(categoryBudka, songZzz, songCma, songJolkaJolka, song51, categoryPinkFloyd, categoryOthers, categoryCustom, songJestTaki, categoryCc), listOf(song51, songCma, songJestTaki, songJolkaJolka, songZzz, categoryCustom, categoryBudka, categoryCc, categoryPinkFloyd, categoryOthers))
    }

    private fun songItem(title: String, categoryType: CategoryType = CategoryType.OTHERS, categoryName: String = "others", custom: Boolean = false): SongTreeItem {
        val category = Category(categoryType.id, categoryType, categoryName, false)
        val song = Song(
                id = "1",
                title = title,
                categories = mutableListOf(category),
                content = null,
                versionNumber = 1,
                createTime = 0,
                updateTime = 0,
                comment = null,
                preferredKey = null,
                locked = false,
                lockPassword = null,
                author = null,
                status = SongStatus.PUBLISHED
        )
        return SongTreeItem.song(song)
    }

    private fun categoryItem(artist: String): SongTreeItem {
        return categoryItem(CategoryType.ARTIST, artist)
    }

    private fun categoryItem(categoryType: CategoryType, categoryName: String): SongTreeItem {
        val category = Category(categoryType.id, categoryType, categoryName, false)
        return SongTreeItem.category(category)
    }

    private fun assertSortedOrdering(`in`: List<SongTreeItem>, outExpected: List<SongTreeItem>) {
        val out = sorter.sort(`in`)
        assertThat(out).isEqualTo(outExpected)
    }

}
